node/src/neighborhood/mod.rs

```diff
@@ -1565,14 +1626,500 @@ mod tests {
+            assert_eq!(
+                actor.overall_connection_status.progress,
+                vec![ConnectionProgress {
+                    initial_node_descriptor: node_descriptor.clone(),
+                    current_peer_addr: new_pass_target,
+                    connection_stage: ConnectionStage::StageZero
+                }]
+            );
+        });
+        let connection_progress_message = ConnectionProgressMessage {
+            peer_addr: node_ip_addr,
+            event: ConnectionProgressEvent::PassGossipReceived(new_pass_target),
+        };
+
+        cpm_recipient.try_send(connection_progress_message).unwrap();
+
+        addr.try_send(AssertionsMessage { assertions }).unwrap();
+        System::current().stop();
+        assert_eq!(system.run(), 0);
+    }
+
+    #[test]
+    fn neighborhood_handles_a_connection_progress_message_with_dead_end_found() {
+        init_test_logging();
+        let cryptde: &dyn CryptDE = main_cryptde();
+        let earning_wallet = make_wallet("earning");
+        let node_ip_addr = IpAddr::from_str("5.4.3.2").unwrap();
+        let node_addr = NodeAddr::new(&node_ip_addr, &[5678]);
+        let this_node_addr = NodeAddr::new(&IpAddr::from_str("1.2.3.4").unwrap(), &[8765]);
+        let public_key = PublicKey::new(&b"booga"[..]);
+        let node_descriptor =
+            NodeDescriptor::from((&public_key, &node_addr, Chain::EthRopsten, cryptde));
+        let mut subject = Neighborhood::new(
+            cryptde,
+            &bc_from_nc_plus(
+                NeighborhoodConfig {
+                    mode: NeighborhoodMode::Standard(
+                        this_node_addr,
+                        vec![node_descriptor.clone()],
+                        rate_pack(100),
+                    ),
+                },
+                earning_wallet.clone(),
+                None,
+                "neighborhood_handles_a_connection_progress_message_with_pass_gossip_received",
+            ),
+        );
+        let (recipient, _) = make_node_to_ui_recipient();
+        subject.node_to_ui_recipient_opt = Some(recipient);
+        subject.overall_connection_status.update_connection_stage(
+            node_ip_addr,
+            ConnectionProgressEvent::TcpConnectionSuccessful,
+            subject.node_to_ui_recipient_opt.as_ref().unwrap(),
+        );
+        let addr = subject.start();
+        let cpm_recipient = addr.clone().recipient();
+        let system = System::new("testing");
+        // let new_pass_target = IpAddr::from_str("10.20.30.40").unwrap();
+        let assertions = Box::new(move |actor: &mut Neighborhood| {
+            assert_eq!(
+                actor.overall_connection_status.progress,
+                vec![ConnectionProgress {
+                    initial_node_descriptor: node_descriptor.clone(),
+                    current_peer_addr: node_ip_addr,
+                    connection_stage: ConnectionStage::Failed(DeadEndFound)
```

Comment:
"DeadEndFound" might be clearer as "PassLoopFound"...or at least something that mentioned pass loops.